File: src/week1/leeson2/trapping-rain-water.cpp
/**
 * 单调栈做法
 */

class Solution {
   public:
    int trap(vector<int>& height) {
        stack<int> s;

        int ans = 0;
        for (int i = 0; i < height.size(); ++i) {
            while (!s.empty() && height[i] > height[s.top()]) {
                int top = s.top();
                s.pop();

                if (s.empty()) break;

                int left = s.top();
                int width = i - left - 1;

                int currentHeight = min(height[i], height[left]) - height[top];

                ans += width * currentHeight;
            }

            s.push(i);
        }

        return ans;
    }
};

File: src/week1/leeson2/trapping-rain-water.ts
export function trap(height: number[]): number {
  const n = height.length
  const leftMax: number[] = Array.from({ length: n })
  leftMax[0] = height[0]
  const rightMax: number[] = Array.from({ length: n })
  for (let i = 1; i < n; ++i) {
    leftMax[i] = Math.max(height[i], leftMax[i - 1])
  }

  rightMax[n - 1] = height[n - 1]
  for (let i = n - 2; i >= 0; --i) {
    rightMax[i] = Math.max(height[i], rightMax[i + 1])
  }

  let ans = 0

  for (let i = 0; i < n; ++i) {
    const min = Math.min(leftMax[i], rightMax[i])

    ans += min <= height[i] ? 0 : min - height[i]
  }

  return ans
}
